File: go/tree/199_right_side_view.go
package tree

func rightSideView(root *TreeNode) []int {
	// 层次遍历 队列先进先出
	// 正常从左到右入队列，现在从右到左

	if root == nil {
		return nil
	}

	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	ans := make([]int, 0)

	for len(queue) > 0 {
		// 第一个元素即为每层最右侧的元素
		ans = append(ans, queue[0].Val)

		for _, v := range queue {
			queue = queue[1:]

			if v.Right != nil {
				queue = append(queue, v.Right)
			}

			if v.Left != nil {
				queue = append(queue, v.Left)
			}
		}
	}
	return ans
}

// rightSideView  层序遍历最右侧的结果
func rightSideView2(root *TreeNode) []int {

	if root == nil {
		return nil
	}

	ans := make([]int, 0)

	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) > 0 {

		// level := make([]int, 0)
		// 优化
		ans = append(ans, queue[len(queue)-1].Val)

		for _, v := range queue {
			queue = queue[1:]
			// level = append(level, v.Val)

			if v.Left != nil {
				queue = append(queue, v.Left)
			}

			if v.Right != nil {
				queue = append(queue, v.Right)
			}

		}

		// ans = append(ans, level[len(level)-1])

	}

	return ans
}

File: rust/src/tree/199_right_side_view.rs


// BFS
pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {

    if root.is_none() {
        return vec![];
    }

    let mut ans = vec![];

    use std::collections::VecDeque;

    let mut queue = VecDeque::new();
    queue.push_back(root);

    while !queue.is_empty() {

        let mut level = vec![];

        for _ in 0..queue.len() {
            
            let head = queue.pop_front().unwrap().unwrap();

            level.push(head.borrow().val);
            
            if head.borrow().left.is_some() {
                queue.push_back(head.borrow_mut().left.take());
            }

            if head.borrow().right.is_some() {
                queue.push_back(head.borrow_mut().right.take());
            }
        }
        // println!("level: {:?}", level);
        ans.push(*level.last().unwrap());
    }

    ans

}